src/awkward/_util.py

```diff
@@ -908,7 +914,7 @@ def visitor(left, right) -> bool:
         elif left.is_option:
             return numpy.array_equal(
                 left.index.data < 0, right.index.data < 0
-            ) and visitor(left.content, right.content)
+            ) and visitor(left.project(), right.project())
```

Comment:
On this line, it looks like you've fixed a potentially subtle bug. Or maybe not that subtle; any test with missing values in it at this level would be sensitive to this difference.